### DISCUSSION TITLE
UI Improvements

### DISCUSSION THREAD

**Discussion by yephny** [TrianguloY/URLCheck#34] [General]
I love the app and and think that it is necessary for any device. Though, I have to point out some of its shortcomings, mainly in its UI, which are evident when compared to its now discontinued alternative [link eye](https://www.f-droid.org/packages/kuesji.link_eye/). As a comparison, I will be linking a screenshot from each app. It could look much cleaner and modern with some improvements.

<img src="https://a.pomf.cat/htgczk.jpg" width=40%> <img src="https://a.pomf.cat/jtopel.jpg" width=40%>

1. The share button is displayed as Sh...ar...e top down.
2. You have to go through one extra tap if you want to open the link with another app in the list. Link eye eliminates this by listing all the compatible apps installed.
3. You either have to select the link or hit the share button to copy the link. An easier way to copy like a designated button could be useful.
4. I haven't tested it with links that have a lot of pattern issues or trackers, so I'm not sure if this makes sense but I think a lot of space could be saved if these sections were side-by-side. They could also be hidden if no issues were found.

I really like the app being an overlay instead of taking up the entire screen. Patterns checker sounds like a great feature and the status code feature is useful especially on social media, where each platform uses its own url shortener.

One feature I did not understand is the custom tabs button. Do you really need it on per-link basis? The toggle could well be placed in the app settings if that's not the case.

**Comment by TrianguloY** [comment]
That link-eye app sure is similar. In fact the list of apps was the original idea of my app! I promise I didn't copied it (although I don't know which app was first published).

As for the other questions:

1) I think it's obvious I'm a backend developer, all my interfaces are ugly and basic, I know. That vertical share button is due to how the layout is positioned, I'll try to set a minimum width.

2) As I have explained, originally the idea was to have the list of apps. However when I changed into the idea of modules there were simply not enough space, so I changed it to a dropdown instead. Hmm, maybe there could be a module that you can toggle to show all apps 🤔

3) Or, you can long tap the share button to directly copy it. Very hidden feature, I know. I should at least explain it in the configuration screen.

4) Side by side modules could be a solution, although that will probably require a custom framework to place and adapt the modules shape. I can try to reduce the minimum height though.

5) I could probably add a setting to enable/disable the custom tabs, or to allow setting it to always enabled/disabled. Personally I use that button when sometimes I need to quickly open an url, from discord for example, or when I prefer to open the whole browser. Not much useful, I agree, but it's just a square button.

**Comment by racuna** [comment]
I like the current UI, maybe it should have an user guide for the first time users: e.g. I don't know what it does the white button in the bottom left.

And maybe it should have some additions to paste some url from the clipboard, an option to check everything without pressing "Check", or adding a QR reader to detect phishing in all the QR's out there (airports, restaurants, and so on).

But after one day checking some URLs from SMS, Whatsapp and Instagram, I really like the core functionality. Thanks you very much.

**Comment by TrianguloY** [comment]
Thanks! As I said designing user interfaces is hard for me, glad you like it.

The modules setting is where each module is explained, I think I forgot to update the description to explain the custom tabs button. When enabled, the browser should be opened in a reduced mode, which theoretically loads faster with less features.

Option to check everything: I made everything that does a network petition or changes the url require a button press on purpose, although some settings to auto apply could be nice.

Pasting from clipboard: that's a strange request, the app opens because you opened a link, if you want to analyze the clipboard link you can open it directly.

QR scanner is similar. You can use any scanner, which should show the link and open it with my app (if it doesn't that's another issue).

**Comment by racuna** [comment]
> Pasting from clipboard: that's a strange request

Ok, I get it. But sometimes there is some urls not highlighted as hyperlinks and could be copied to the clipboard.

> You can use any scanner

Yep, you are all right. Thanks

**Comment by TrianguloY** [comment]
> sometimes there is some urls not highlighted as hyperlinks and could be copied to the clipboard.

Hmm, can you specify what you mean with steps or similar? I'm sorry but I don't really understand what you mean :/

**Comment by racuna** [comment]
> Hmm, can you specify what you mean with steps or similar? I'm sorry but I don't really understand what you mean :/

when an url is "not clickable", it its just text, but you can select it an copy it to the clipboard, then paste it to check it

a rare case of use, tho.

in other topic: I open an ad on Google Search app, appears a humongous URL, that provoked an internal error on the  "url checker" app. But I just was testing, I don't really click on ads in real life situation :P

**Comment by TrianguloY** [comment]
Ah! not rare at all. Select the text link, but instead of pressing 'copy' press 'share'. My app should appear in the list (it is registered as capable of handling plain text), choose it and the expected screen should open.

Internal crash? That's not good. Could it be possible to share the ad link here or via email? I can take a look.

**Comment by racuna** [comment]
> Select the text link, but instead of pressing 'copy' press 'share'
tested, it works, thanks.

I opened Google App, the ad was different this time but the URL was still giant AF:

Screenshot of the error: https://i.imgur.com/9LlzmGd.jpg

humongous URL will send it to you by email because pastebin didn't accept it

**Comment by yephny** [comment]
Your entire exchange could be transferred to a discussions page or another issue as there doesn't seem to be anything pertaining to the UI.

**Comment by TrianguloY** [comment]
Some of the replies can be useful for the issue I think, but I agree it's mostly off topic.

I can't seem to move individual comments, it's the whole issue or nothing, so I'll leave this as issue unless you prefer to move it, and I'll try to remember to reply as conversation in the future.

**Comment by ig9000** [comment]
One UI feature that I've seen elsewhere that would relate to this is to display direct buttons for the last/most used app from the dropdown list. Two buttons, one for last used and one for most used, would cover majority of use cases for me.

**Comment by TrianguloY** [comment]
That's a nice idea, maybe it is better to move to a new feature issue, to be honest the 'open module' needs a big rework.

**Comment by yephny** [comment]
I really like having all compatible apps listed, though I don't know how it could be implemented in Url checker.

**Comment by Ilithy** [comment]
For my part I find the interface successful, the app supports monet, which is very pretty, the only regret is the "Open with" button, which in my opinion could be almost identical to the Share button next to it (or with a thin rounded frame, and a small shade of color).

![Screenshot_20220724-191751](https://user-images.githubusercontent.com/36798218/180659082-5e7876ae-f47e-4b7e-8e45-9abebcb1671a.png)

**Comment by TrianguloY** [comment]
Thanks, although I don't know what is "monet" 😅

And to be honest I also dislike the open button, but when I changed all of them from the default (that one) to the borderless (the others) it was hard for me to find the open button, I really disliked it, so I kept it this way for emphasis.

The thing is that I'm using the device theme, so on newer androids it uses the material theme, on older the holo, and hopefully on future devices the [awesome theme name] theme. I do this to avoid extra assets (most apps size are almost always visual things, not code) but android's theming is horrible, no wonder companies create their own... The main issue is that if I set a very personalized button, it may look really bad on [awesome device] models. I'll check if I can at least tweak it a bit though.

**Comment by Ilithy** [comment]
I understand, these are constraints that a basic user does not necessarily understand (like me).

As for monet, I believe that it is more generally called material3, UrlChecker supports it, because it modifies the colors (of accent) according to the wallpaper (mine and on orange tone, for example ), it's very nice.

this button is not a horror either ^^
It's just the least graceful element of the interface.

FYI this rendering is that of AOSP 12L (I run under the latest stable version of Grapheneos), with the colors adapted to my background, of course :-)

**Comment by TrianguloY** [comment]
> I understand, these are constraints that a basic user does not necessarily understand (like me).

Think of it like "This is a standard/borderless button" (what my app does, let's the device choose how to display it, just an couple lines but each android/device shows it different) vs "This is a lightgrey rectangle with a red 2 pixel wide border with rounded corners and a text with size 15 color black and..." (what other apps do, exactly how they want but very resource heavy).

> As for monet, I believe that it is more generally called material3, UrlChecker supports it, because it modifies the colors (of accent) according to the wallpaper (mine and on orange tone, for example ), it's very nice.

Ah! Yes, that's one of the benefits of letting the device choose the theme.

> this button is not a horror either ^^
> It's just the least graceful element of the interface.
> 
> FYI this rendering is that of AOSP 12L (I run under the latest stable version of Grapheneos), with the colors adapted to my background, of course :-)

Looks good, just like my emulator! (My daily phone has android 12 but not the pixel-only features, sadly)

**Comment by TrianguloY** [comment]
Honest question: What do you (anyone reading this) think about this?
I think it's too much color, but I really love the new open button. And I managed to make almost everything that is clickable colored with the device accent color, and everything that isn't black&white; so it's consistent.

![image](https://user-images.githubusercontent.com/10616202/181498315-7224fce8-4c42-4dca-9c38-0a863b75566d.png)
![image](https://user-images.githubusercontent.com/10616202/181498619-8c1f228c-86b9-49bc-84f8-0c28f4e81e43.png)

**Reply by Ilithy** [reply]
>I think it's too much color, but I really love the new open button.

I do not find, on the contrary, it attracts attention, and with the dynamic colors, it is very pleasant.

>And I managed to make almost everything that is clickable colored with the device accent color, and everything that isn't black&white; so it's consistent.

Yes, it gives a fairly clear consistency, the accent palette is well managed (on my smartphone I only have criart backgrounds, and it looks very very good, while remaining readable), I have nothing to complain about ( and yet I'm looking ^^)

Thank you very much for your work

**Comment by racuna** [comment]
I like the colors. How it will look like when there is more than one app for the same kind of link?
e.g. youtube video: Youtube app and Newpipe

**Comment by TrianguloY** [comment]
![image](https://user-images.githubusercontent.com/10616202/181710894-83e90496-8f13-4d2a-8646-4fec63aa9066.png)

And the dropdown is the same, that's a built-in android component.

**Comment by Ilithy** [comment]
Really sorry for the delay, I was not notified of your answers.

I personally think it's very pretty, it sticks very well to the rest and identically for the drop-down menu, it comes out well, clearly it has style! <3

**Comment by ig9000** [comment]
> That's a nice idea, maybe it is better to move to a new feature issue,

Sorry, I never got around to this, but now that this thread has been converted to a "discussion", hopefully it's ok to toss out ideas that aren't fully-formed enough to warrant an issue?

I like the mock-ups. As a single data point of feedback: As long as an app avoids forcing some arbitrary accent color (or even worse, a white background), I find the more subtle details of widget rendering to not impact my experience much. Geometry does, though, and on that note, I'll also say that I'd love to see the share menu given at least as much real estate as the browser menu. This is because my use of URLChecker (and predecessors) is in part as a workaround for apps (eg. QKSMS) which fail to provide a share function for URLs. I probably use the share function about twice as often as the browser button, for whatever that's worth.

That said, another little enhancement would be a preference setting for having the app close itself after selecting from the menu. I personally almost never need to interact with it again after that. Just a thought. In any case, thanks for your work on this!

**Reply by TrianguloY** [reply]
The color is the accent color of the device. On android 12 some devices allow you to change it from the settings, and I think since android 8 there are some ways to change it from adb or root. In any other case I'm afraid the color is "forced" but at least is your device color.
Having an option to change it from settings could be nice, I would need to investigate how to do so though.

Your share usage is interesting, the main feature of the app is the open button, yes,  and the share button is just an extra option there. From another issue there has been some discussion about splitting the open and share functions into different modules. This will separate them (one above the other, reorderable) and will allow for more "features" of each functionality.
The share module could have a separated "copy to clipboard" button (instead of the current long click), can you think of any other functionality that could be added? 

Yes, an option to auto close the app is a good idea, can you write a new issue with it?

**Reply by ig9000** [reply]
> From another issue there has been some discussion about splitting the open and share functions into different modules.

That sounds like a better idea than having them compete for space on a single row, since I personally pass URLs to non-browser apps quite often. I can't think of other functionality to add at the moment, but I'll ponder it.

> Yes, an option to auto close the app is a good idea, can you write a new issue with it?

Done: #42

**Comment by Ilithy** [comment]
If you remove the "copy to clipboard" functionality from the share button (to make it a standalone option), please could you also add an option to merge them back into one share button?

In additional ideas, the history module could be more powerful, by keeping the links (before and after processing) passing through it, and allowing the export (and import) of this history base (the number of links kept and/or the weight of this base could be configurable, eventually).
Is this possible?

many thanks

**Reply by Ilithy** [reply]
On another subject (but still aesthetic), some denominations could be more harmonious in my opinion, See:

Line 97  :    `<string name="mVT_name">Scan url</string>` in `<string name="mVT_name">Url Scanner</string>`  
Line 120:    `<string name="mClear_name">Clear url</string>` in `<string name="mClear_name">Url Cleaner</string>`  
Line 153:    `<string name="mRemove_name">Remove Queries</string>` in `<string name="mRemove_name">Queries Remover</string>`

I find this more consistent, and it follows the pattern of line 35:
`<string name="mPttrn_name">Patterns checker</string>`

**Reply by TrianguloY** [reply]
You prefer to long tap instead of a dedicated button? Or you mean to keep the share and open buttons in the same module?



The history keeps all links that pass, except when writing in the textview (to avoid having "g", "go", "goo", "googl"...). Also when you go back and set another link the previous "newer" are discarded.
It's the standard functionality, what I mean is: you open A, then change to B and C. If you go back to A and change to D, the history will contain A>D. It is be possible to keep all, but why?
And why do you want to export or import the list?

Wait...you mean to keep all opened links from each time you open the app? 


Those string changes are a very good idea! Do you want to implement them yourself? I can do it easily this afternoon if not.

**Reply by Ilithy** [reply]
> You prefer to long tap instead of a dedicated button? Or you mean to keep the share and open buttons in the same module?

Honestly I like how it is now, and how it works so both :-),
![Capture](https://user-images.githubusercontent.com/36798218/183076435-0555349b-48b8-44a0-9e4f-c40cd95b75cd.PNG)
_**You admit it, it's sleek, beautiful, clean, and practical!**_ <3

I understand that another method with more buttons would offer more functionality, but I like this design ;-)
But I don't want to stop you from improving Url Checker.

>Wait...you mean to keep all opened links from each time you open the app?

Yes, that's it, in part, but I'll go a little further...

1. It will only keep in its database the A versions (and possibly the D version too), the B and C being useless in this case.
2. The History module would have a new button to add the url to the url database.
3. Would have an option to automatically add all links passing through Url Checker to the DB.
4. Allow the export (and import) of the url database
5. Allow to add a short comment on the urls contained in the DB)

_This is probably too complex, but hope is free_

The interest of the export (but also intrinsically of the DB), it would be especially to make of Url a replacement (multitache) to keep favorite URLs (and also to make of it a curator of very diverse bookmarks, that would be extra)

>Those string changes are a very good idea! Do you want to implement them yourself? I can do it easily this afternoon if not.

Thanks, its done

**Comment by Ilithy** [comment]
rendering on my Pixel 6 Pro is awesome. 🤩
![Screenshot_20220813-200635__01__01](https://user-images.githubusercontent.com/36798218/184506189-c5ce9d8e-f5f9-45c3-a7de-3e9dce46c601.png)

A question (not directly related): would it be possible to make the enclosed text translatable?

Thanks a lot.

**Reply by TrianguloY** [reply]
Glad you like it :)
Since then I managed to fix the padding issues (I hope) and increased slightly the button's minimum height (personally I found it too small). Next beta should appear soon! (If not I'll just make a new alpha)


Wait, the button descriptions are not translatable? 🤦 My bad, I'll extract them next.

**Reply by Ilithy** [reply]
>increased slightly the button's minimum height (personally I found it too small)

I don't find, but I can be totally wrong

Thanks for all 🙏
